stdlib/LinearAlgebra/src/bunchkaufman.jl

```diff
@@ -390,3 +411,1225 @@ end
 ## reconstruct the original matrix
 ## TODO: understand the procedure described at
 ## http://www.nag.com/numeric/FL/nagdoc_fl22/pdf/F07/f07mdf.pdf
+
+
+##--------------------------------------------------------------------------
+##------------- Start of generic Bunch-Kaufman Implementation --------------
+##--------------------------------------------------------------------------
+
+export inertia
+
+function arg_illegal(fun_name::AbstractString,
+    info::Integer,
+    waer::AbstractChar)
+    if waer == 'W'
+        @warn " ** On entry to '$(fun_name)' parameter number " *
+            "$(info) had an illegal value"
+    else
+        error(" ** On entry to '$(fun_name)' parameter number " *
+            "$(info) had an illegal value")
+    end
+end
+
+
+function cabs1(z::T) where T <: Complex
+    return abs(real(z)) + abs(imag(z))
+end
+
+
+function cabsr(z::T) where T <: Complex
+    return abs(real(z))
+end
+
+
+"""
+generic_adr1!(uplo, alpha, x, y, A, syhe) -> nothing
+
+`generic_adr1!` performs the following adjoint (symmetric or Hermitian)
+rank 1 operation
+
+`A[1:K,1:L] = alpha*x*y' + A[1:K,1:L]`
+
+in-place, where `alpha` is a scalar, `x` is a K element vector, `y`
+is an L element vector and `A` is an `NxM` matrix. Note that `y'` can
+denote either the transpose, i.e. `transpose(y)` or the conjugate
+transpose , i.e. `adjoint(y)`.
+
+`uplo` is a character, either `'U'`, `'L'` or `'F'`, indicating whether
+the matrix is stored in the upper triangular part (`uplo=='U'`), the
+lower triangular part (`uplo=='L'`), or the full storage space is used
+(`uplo=='F'`). If `uplo!='F'` then only the corresponding triangular
+part is updated. The values `'U'` or `'L'` can onlt be used when A is
+square (`N==M`).
+
+`syhe` is a character, either `'S'` or `'H'`, indicating whether the
+symmetric adjoint (`syhe=='S'`, and `y'==tranpose(y)`) or the hermitian
+adjoint (`syhe=='H'`, and `y'==adjoint(y)`) must be used.
+"""
+function generic_adr1!(uplo::AbstractChar,
```

Comment:
rather than following a BLAS interface here use types (i.e. no `uplo` or `syhe`)

Reply:
I guess you something like `if A.uplo == :U ...`, am I right?

Reply:
just `if A isa UpperTriangular`/`if A isa LowerTriangular`. Similarly, `syhe` is just `if A isa Hermitian`/`if A isa Symmetric`

Reply:
The code performs the factorization in place, so `A` in general will be `Symmetric` or `Hermitian`. The `uplo` check is done to know which triangular part of `A` stores the data. Comparing with `UpperTriangular` or `Lower Triangular` is not feasible in this case.

Reply:
I have made the changes where in most functions, except for the main factorization function `bunchkaufman_native!` and its helper functions `generic_adr1!` and `bk_rowcol_swap!`. LAPACK works with the raw data directly and so the `BunchKaufman` object stores the raw data as a generic matrix. The main factorization function works with the raw data, and modifying it to work with `Symmetric`, `UpperTriangular`, etc. wrappers only to extract the raw data to store it in the `BumchKaufman` object feels a bit redundant. Of course, I modified the rest of the functions to work directly with the `BunchKaufman` object if available, so the code has simplified a bit. Is this acceptable?